File: src/voice_commands.py
"""
Voice command post-processing for transcribed text.

Replaces spoken formatting commands (e.g. "new line", "period", "comma")
with their corresponding characters. Applied after transcription, before
paste or save. Each command can be individually enabled or disabled.
"""

from __future__ import annotations

import logging
import re
from typing import Dict

logger = logging.getLogger(__name__)

# All available voice commands: (spoken phrase, replacement character).
# Ordered longest-first to avoid partial matches
# (e.g. "new paragraph" must be matched before "new").
ALL_COMMANDS: list[tuple[str, str]] = [
    ("new paragraph", "\n\n"),
    ("new line", "\n"),
    ("exclamation point", "!"),
    ("exclamation mark", "!"),
    ("question mark", "?"),
    ("open parenthesis", "("),
    ("close parenthesis", ")"),
    ("open quote", '"'),
    ("close quote", '"'),
    ("semicolon", ";"),
    ("period", "."),
    ("comma", ","),
    ("colon", ":"),
]

# Display-friendly names for the replacement characters.
REPLACEMENT_DISPLAY: dict[str, str] = {
    "\n\n": "\\n\\n",
    "\n": "\\n",
    "!": "!",
    "?": "?",
    "(": "(",
    ")": ")",
    '"': '"',
    ";": ";",
    ".": ".",
    ",": ",",
    ":": ":",
}

# Punctuation characters that should not have a leading space.
_PUNCT_NO_LEADING_SPACE = re.compile(r"\s+([.,!?;:\)\"])")
# Opening brackets/quotes should not have a trailing space.
_PUNCT_NO_TRAILING_SPACE = re.compile(r"([\(\"])\s+")


def default_enabled() -> Dict[str, bool]:
    """Return the default enabled state for all commands (all True)."""
    return {cmd: True for cmd, _ in ALL_COMMANDS}


def apply_voice_commands(text: str, enabled: Dict[str, bool] | None = None) -> str:
    """Replace spoken voice commands with formatting characters.

    Parameters
    ----------
    text:
        Raw transcribed text that may contain spoken commands like
        "new line", "period", "comma", etc.
    enabled:
        Dict mapping command phrase to bool. Commands not in the dict
        or set to False are skipped. If None, all commands are applied.

    Returns
    -------
    str
        Text with voice commands replaced by their characters and
        whitespace cleaned up around punctuation.
    """
    if not text:
        return text

    count = 0
    result = text

    for cmd, replacement in ALL_COMMANDS:
        if enabled is not None and not enabled.get(cmd, False):
            continue
        # Also consume trailing punctuation the model adds after the command
        # (e.g. "New line." → the trailing "." is an auto-punctuation artifact).
        pattern = re.compile(
            r"\b" + re.escape(cmd) + r"\b[.,;:!?]*", re.IGNORECASE
        )
        result, n = pattern.subn(replacement, result)
        count += n

    if count > 0:
        # Clean up spaces around newlines (e.g. "Hello \n world" → "Hello\nworld").
        result = re.sub(r" *\n *", "\n", result)
        # Remove orphaned punctuation at start of lines (model artifacts).
        result = re.sub(r"\n[.,;:!?]+\s*", "\n", result)
        # Clean up whitespace around punctuation.
        result = _PUNCT_NO_LEADING_SPACE.sub(r"\1", result)
        result = _PUNCT_NO_TRAILING_SPACE.sub(r"\1", result)
        # Collapse multiple spaces into one.
        result = re.sub(r" {2,}", " ", result)
        logger.info("Voice commands: %d replacement(s) applied", count)

    return result

File: ParkeetSwift/Sources/Parkeet/Model/VoiceCommands.swift
import Foundation

/// Post-processes transcribed text, replacing spoken formatting commands with symbols.
///
/// Example: "Hello new line world period" → "Hello\nworld."
enum VoiceCommands {

    // Commands ordered longest-first (greedy matching)
    static let commands: [(phrase: String, replacement: String)] = [
        ("new paragraph", "\n\n"),
        ("new line", "\n"),
        ("exclamation point", "!"),
        ("exclamation mark", "!"),
        ("question mark", "?"),
        ("open parenthesis", "("),
        ("close parenthesis", ")"),
        ("open quote", "\""),
        ("close quote", "\""),
        ("semicolon", ";"),
        ("period", "."),
        ("comma", ","),
        ("colon", ":"),
        ("dash", "—"),
        ("hyphen", "-"),
    ]

    /// Default enabled state for each command.
    static let defaultEnabled: [String: Bool] = {
        var dict: [String: Bool] = [:]
        for cmd in commands {
            dict[cmd.phrase] = true
        }
        return dict
    }()

    /// Apply enabled voice commands to transcribed text.
    static func apply(to text: String, enabled: [String: Bool]) -> String {
        var result = text

        // Replace each enabled command phrase with its symbol
        for cmd in commands {
            guard enabled[cmd.phrase] ?? true else { continue }

            // Case-insensitive replacement
            result = result.replacingOccurrences(
                of: cmd.phrase,
                with: cmd.replacement,
                options: .caseInsensitive
            )
        }

        // Clean up whitespace around punctuation
        result = cleanupWhitespace(result)

        return result.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Remove spaces before punctuation, after opening quotes/parens.
    private static func cleanupWhitespace(_ text: String) -> String {
        var result = text

        // Remove space before punctuation: "Hello ." → "Hello."
        let punctuation = [".", ",", "!", "?", ";", ":", "—", "-", ")", "\""]
        for p in punctuation {
            result = result.replacingOccurrences(of: " \(p)", with: p)
        }

        // Remove space after opening delimiters: "( hello" → "(hello"
        result = result.replacingOccurrences(of: "( ", with: "(")
        result = result.replacingOccurrences(of: "\" ", with: "\"")

        // Collapse multiple spaces
        while result.contains("  ") {
            result = result.replacingOccurrences(of: "  ", with: " ")
        }

        // Remove orphaned punctuation at line starts (model artifacts)
        let lines = result.components(separatedBy: "\n")
        let cleaned = lines.map { line -> String in
            var l = line.trimmingCharacters(in: .whitespaces)
            while let first = l.first, ".!?,;:".contains(first) {
                l = String(l.dropFirst()).trimmingCharacters(in: .whitespaces)
            }
            return l
        }
        result = cleaned.joined(separator: "\n")

        return result
    }
}
